File: sky_news_demo/operations/mediaconvert/getMediaConvert.js
var AWS = require("aws-sdk");

/**
 * Get status of an Amazon Media Convert job
 */

exports.handler = async (event) => {
    
    console.log('[INFO] got event: %j', event);
    
    try
    {
        var endpointParams = await getMediaConvertEndpoint();
        var mediaConvertOutput = await getMediaConvertStatus(event, endpointParams);
        console.log(mediaConvertOutput);

        if (mediaConvertOutput == 'Executing')
        {
            console.log(event);
            return event;
        }
        if (mediaConvertOutput == 'Error')
        {
            let output = {'name': 'mediaconvert', 'status': 'Error', 'metadata': {'job_id': event.metadata.job_id, 'input_key': event.metadata.input_key} };
            return output;
        }
        else
        {
            let output = {'name': 'mediaconvert', 'status': 'Complete', 'media': {'audio': {'s3bucket': mediaConvertOutput.s3bucket, 's3key': mediaConvertOutput.s3key}}, 'metadata': {'job_id': event.metadata.job_id, 'input_key': event.metadata.input_key} };
            return output;
        }

    }
    catch (error)
    {
        console.log('[ERROR] failed to get status of MediaConvert job', error);
        throw error;
    }
};

async function getMediaConvertEndpoint() {

         // Instantiate mediaconvert
        try
        {
        let mediaconvert = new AWS.MediaConvert({
            region: process.env.AWS_REGION
          });

        // Get mediaconvert endpoint

        var endpoints = await mediaconvert.describeEndpoints().promise();
        var params = {
            "endpoint": endpoints.Endpoints[0].Url,
            "region": process.env.AWS_REGION
             };
        console.log(params);
        return params;
        }
        catch(error)
        {
            console.log("[ERROR] failed to get MediaConvert Endpoint", error);
		    throw error;
        }
}

async function getMediaConvertStatus(event, endpointParams)
{
    console.log(event);
    try
    {

       // Get setup new mediaconvert object with params from getMediaConvertEndpoint

        let mediaconvert = new AWS.MediaConvert({
            endpoint: endpointParams.endpoint,
            region: endpointParams.region
          });

       var id = event.metadata.job_id;

       var params = {
            Id: id
          };

       var response = await mediaconvert.getJob(params).promise();

       console.log(response);

       if (response.Job.Status == 'IN_PROGRESS' || response.Job.Status == 'PROGRESSING')
       {
          return 'Executing';
       }

       if (response.Job.Status == 'COMPLETE')
       {
          var uri = response.Job.Settings.OutputGroups[0].OutputGroupSettings.FileGroupSettings.Destination;

          console.log(uri);
          console.log(response.Job.Settings.OutputGroups[0].OutputGroupSettings.FileGroupSettings);
          console.log(response.Job.Settings.OutputGroups[0].OutputGroupSettings);
          console.log(response.Job.Settings.OutputGroups[0]);
          console.log(response.Job.Settings.OutputGroups);

          // Need to clean this up and find a better way to get the output key

          var split_uri = uri.split("/");
          var bucket = split_uri[2];
          var folder = split_uri[3];

          var input_file = event.metadata.input_key;
          var input_file_split = input_file.split("/");
          var file_name = input_file_split[1];
          var file_name_split = file_name.split(".");
          var name = file_name_split[0];

          var extension = response.Job.Settings.OutputGroups[0].Outputs[0].Extension;
          var modifier = response.Job.Settings.OutputGroups[0].Outputs[0].NameModifier;


          var key = folder + "/" + name + modifier + "." + extension;

          return {'s3bucket': bucket, 's3key': key};
          }
       if (response.Job.Status == 'ERROR')
       {
          return 'Error';
       }

    }
    catch(error)
    {
        console.log("[ERROR] failed to get status of MediaConvert job", error);
		throw error;
    }
}

File: sky_news_demo/operations/mediaconvert/get_media_convert.py
import os
import boto3
from outputHelper import OutputHelper
from outputHelper import MasExecutionError

region = os.environ["AWS_REGION"]

mediaconvert = boto3.client("mediaconvert", region_name=region)

operator_name = "mediaconvert"
output_object = OutputHelper(operator_name)

def lambda_handler(event, context):
    print("We got the following event:\n", event)

    job_id = event["metadata"]["mediaconvert_job_id"]

    try:
        response = mediaconvert.describe_endpoints()
    except Exception as e:
        print("Exception:\n", e)
        output_object.update_status("Error")
        output_object.update_metadata(mediaconvert_error=str(e))
        raise MasExecutionError(output_object.return_output_object())
    else:
        mediaconvert_endpoint = response["Endpoints"][0]["Url"]
        customer_mediaconvert = boto3.client("mediaconvert", region_name=region, endpoint_url=mediaconvert_endpoint)

    try:
        response = customer_mediaconvert.get_job(Id=job_id)
    except Exception as e:
        print("Exception:\n", e)
        output_object.update_status("Error")
        output_object.update_metadata(mediaconvert_error=e, mediaconvert_job_id=job_id)
        raise MasExecutionError(output_object.return_output_object())
    else:
        if response["Job"]["Status"] == 'IN_PROGRESS' or response["Job"]["Status"] == 'PROGRESSING':
            output_object.update_status("Executing")
            output_object.update_metadata(mediaconvert_job_id=job_id, mediaconvert_input_file=event["metadata"]["mediaconvert_input_file"])
            return output_object.return_output_object()
        elif response["Job"]["Status"] == 'COMPLETE':
            output_uri = response["Job"]["Settings"]["OutputGroups"][0]["OutputGroupSettings"]["FileGroupSettings"]["Destination"]

            extension = response["Job"]["Settings"]["OutputGroups"][0]["Outputs"][0]["Extension"]
            modifier = response["Job"]["Settings"]["OutputGroups"][0]["Outputs"][0]["NameModifier"]

            bucket = output_uri.split("/")[2]
            folder = output_uri.split("/")[3]

            file_name = event["metadata"]["mediaconvert_input_file"].split("/")[1].split(".")[0]

            key = folder + "/" + file_name + modifier + "." + extension
            output_object.update_media("audio", bucket, key)
            output_object.update_metadata(mediaconvert_job_id=job_id)
            output_object.update_status("Complete")

            return output_object.return_output_object()




